src/typed-ast/print.cc

```diff
@@ -0,0 +1,250 @@
+
+#include "typed-ast/ast.h"
+
+#include <iomanip>
+#include <vector>
+
+namespace verona::ast
+{
+  namespace
+  {
+    /// Custom ostreambuf which counts the number of outputted characted.
+    /// The actual contents printed to the stream are discarded.
+    struct length_ostreambuf : public std::streambuf
+    {
+    public:
+      size_t size() const
+      {
+        return size_;
+      }
+
+    protected:
+      std::streamsize xsputn(const char*, std::streamsize n) override
+      {
+        size_ += n;
+        return n;
+      };
+
+    private:
+      size_t size_ = 0;
+    };
+
+    /// Print a quoted string to `out`, escaping any special characters.
+    // TODO: This is a very primitive implementation, which almost certainly
+    // doesn't handle non-ASCII character correctly. It's good enough for
+    // printing AST dumps.
+    void print_string(std::ostream& out, std::string_view s)
+    {
+      out << '"';
+      for (char c : s)
+      {
+        if (c == '\\' || c == '"')
+          out << '\\' << c;
+        else if (c == '\a')
+          out << "\\a";
+        else if (c == '\b')
+          out << "\\b";
+        else if (c == '\e')
+          out << "\\e";
+        else if (c == '\f')
+          out << "\\f";
+        else if (c == '\n')
+          out << "\\n";
+        else if (c == '\r')
+          out << "\\r";
+        else if (c == '\t')
+          out << "\\t";
+        else if (c == '\v')
+          out << "\\v";
+        else if (c == '\0')
+          out << "\\0";
+        else if (c >= 0x20 && c < 0x7f)
+          out << c;
+        else
+        {
+          unsigned int code =
+            static_cast<unsigned int>(static_cast<unsigned char>(c));
+          std::ios_base::fmtflags f(out.flags());
+          out << std::hex << std::setw(2) << std::setfill('0');
+          out << "\\x" << code;
+          out.flags(f);
+        }
+      }
+      out << '"';
+    }
+
+    void print_compact(std::ostream& out, const Node& node)
+    {
+      NodePrinter printer(out, true, 0, 0);
+      node.print(printer);
+    }
+
+    void print_extended(
+      std::ostream& out, const Node& node, size_t indent, size_t width)
+    {
+      NodePrinter printer(out, false, indent, width);
+      node.print(printer);
+    }
+
+    /// Compute the length that would be occupied if the node was printed in its
+    /// compact form.
+    ///
+    /// We do this by running the actual printing code, but on a special ostream
+    /// the counts the number of printed characters.
+    size_t compact_length(const Node& node)
+    {
+      length_ostreambuf buf;
+      std::ostream out(&buf);
+      print_compact(out, node);
+      return buf.size();
+    }
+  }
+
+  void print(std::ostream& out, const Node& node, size_t indent, size_t width)
+  {
+    // We first try to print the node in its compact form. If this takes more
+    // space than what is available, we fall back to the extended form.
+    size_t len = compact_length(node);
```

Comment:
I worry about the quadratic blow up here on printing large ASTs.  But fine for a first cut, and I guess we will only ever use this for testing.  It seems you are really calculating `compact_length` for each node, adding a memoization to that would make things much better from a complexity point of view.

Reply:
Yeah, I wonder how bad the blow up is. The idea was taken from [Pony's own pretty-printer](https://github.com/ponylang/ponyc/blob/eb6fbbb16600249e1d90cd3e6674ebe75fecc5b0/src/libponyc/ast/ast.c) which probably has the same issue. The pony one doesn't use the fake ostream trick to compute the length, but instead has an explicit computation, which is probably faster but more fragile too. The complexity would be the same in both cases though.

Reply:
Added a note explaining the risk, and pointed to memoization as a possible solution.